dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java

```diff
@@ -104,64 +109,79 @@ public HibernatePotentialDuplicateStore( SessionFactory sessionFactory, JdbcTemp
     {
-        String queryString = "from PotentialDuplicate pr where pr.status in (:status)";
+        CriteriaBuilder cb = getCriteriaBuilder();
 
-        return Optional.ofNullable( query.getTeis() ).filter( teis -> !teis.isEmpty() ).map( teis -> {
-            Query<PotentialDuplicate> hibernateQuery = getTypedQuery(
-                queryString + " and ( pr.original in (:uids) or pr.duplicate in (:uids) )" );
+        CriteriaQuery<PotentialDuplicate> cq = cb
+            .createQuery( PotentialDuplicate.class );
 
-            hibernateQuery.setParameterList( "uids", teis );
+        Root<PotentialDuplicate> root = cq.from( PotentialDuplicate.class );
 
-            setStatusParameter( query.getStatus(), hibernateQuery );
+        cq.where( getQueryPredicates( criteria, cb, root ) );
 
-            return hibernateQuery.getResultList();
-        } ).orElseGet( () -> {
+        cq.orderBy( criteria.getOrder().stream().map( order -> order.getDirection()
+            .isAscending() ? cb.asc( root.get( order.getField() ) )
+                : cb.desc( root.get( order.getField() ) ) )
+            .collect( Collectors.toList() ) );
 
-            Query<PotentialDuplicate> hibernateQuery = getTypedQuery( queryString );
+        TypedQuery<PotentialDuplicate> relationshipTypedQuery = getSession()
+            .createQuery( cq );
 
-            setStatusParameter( query.getStatus(), hibernateQuery );
+        if ( criteria.isPagingRequest() )
+        {
+            relationshipTypedQuery.setFirstResult( criteria.getFirstResult() );
+            relationshipTypedQuery.setMaxResults( criteria.getPageSize() );
+        }
 
-            return hibernateQuery.getResultList();
-        } );
+        return relationshipTypedQuery.getResultList();
     }
 
-    private void setStatusParameter( DeduplicationStatus status, Query<?> hibernateQuery )
+    private Predicate[] getQueryPredicates( PotentialDuplicateCriteria query, CriteriaBuilder builder,
+        Root<PotentialDuplicate> root )
     {
-        if ( status == DeduplicationStatus.ALL )
-        {
-            hibernateQuery.setParameterList( "status", Arrays.stream( DeduplicationStatus.values() )
-                .filter( s -> s != DeduplicationStatus.ALL ).collect( Collectors.toSet() ) );
-        }
-        else
+        List<Predicate> predicateList = new ArrayList<>();
+
+        predicateList.add( root.get( "status" ).in( getInStatusValue( query.getStatus() ) ) );
+
+        if ( !query.getTeis().isEmpty() )
         {
-            hibernateQuery.setParameterList( "status", Collections.singletonList( status ) );
+            predicateList.add( builder.and(
+                builder.or(
+                    root.get( "original" ).in( query.getTeis() ),
+                    root.get( "duplicate" ).in( query.getTeis() ) ) ) );
         }
+
+        Predicate[] predicates = new Predicate[predicateList.size()];
```

Comment:
Is intellij flagging this for you? 🤔 

According to https://stackoverflow.com/questions/9572795/convert-list-to-array-in-java

`predicateList.toArray(new Predicate[0]);`

should be preferred